File: python/y2019/d01/aoc_2019_01.py
from common import timer, Puzzle


class Day01(Puzzle, year=2019, day=1):
    @staticmethod
    def clean_input(data: list[str]) -> list[int]:
        return [int(d) for d in data]

    @staticmethod
    def calc_fuel(mass) -> int:
        fuel: int = mass // 3 - 2
        if fuel < 0:
            return 0
        else:
            return fuel + Day01.calc_fuel(fuel)

    @timer(part=1)
    def solve_part_1(self) -> int:
        """Solution for part 1"""
        self.data: list[int]
        return sum([(m // 3 - 2) for m in self.data])

    @timer(part=2)
    def solve_part_2(self) -> int:
        """Solution for part 2"""
        sum_fuel: int = 0
        for mass in self.data:
            fuel: int = mass // 3 - 2
            sum_fuel += fuel + Day01.calc_fuel(fuel)

        return sum_fuel

File: C++/2019/aoc_2019_01.cpp
#include "aoc_io.hpp"
#include "aoc_timer.hpp"
#include <numeric>
#include <ranges>
#include <string>
#include <vector>

static int fuel_cost(int fuel) {
    fuel = (fuel / 3) - 2;
    if (fuel < 0) {
        return 0;
    }
    return fuel + fuel_cost(fuel);
}

std::vector<int> clean_input(const std::vector<std::string> &instructions) {
    std::vector<int> return_instructions;

    return_instructions.reserve(instructions.size());

    for (auto &row : instructions) {
        return_instructions.push_back(std::stoi(row));
    }

    return return_instructions;
}

int solve_part_1(const std::vector<int> &instructions) {

    auto fuel = instructions | std::views::transform([](const int &mass) -> int { return mass / 3 - 2; });

    auto result = std::reduce(fuel.begin(), fuel.end());
    return result;
}

int solve_part_2(const std::vector<int> &instructions) {

    auto fuel = instructions | std::views::transform([](const int &mass) -> int { return fuel_cost(mass); });

    auto result = std::reduce(fuel.begin(), fuel.end());
    return result;
}

void solve_all(const std::vector<std::string> &instructions) {
    std::vector<int> parsed = clean_input(instructions);
    aoc::timer(1, solve_part_1, parsed);
    aoc::timer(2, solve_part_2, parsed);
}

int main(int argc, char **argv) {
    std::string   filename;
    constexpr int year = 2019;
    constexpr int day  = 1;

    std::vector<std::string> instructions;

    if (argc > 1) {
        if (std::string(argv[1]) == "--test") {
            filename = "test_input.txt";
        } else {
            filename = argv[1];
        }
    } else {
        filename = "input.txt";
    }

    instructions = aoc::io::get_input_list<std::string>(filename, year, day);

    aoc::io::header(year, day);
    aoc::timer(0, solve_all, instructions, false);

    return 0;
}
